src/core/server/logging/logging_config.ts

```diff
@@ -62,20 +58,19 @@ const levelSchema = schema.oneOf(
  */
 export const loggerSchema = schema.object({
   appenders: schema.arrayOf(schema.string(), { defaultValue: [] }),
-  context: schema.string(),
   level: levelSchema,
 });
-
 /** @public */
 export type LoggerConfigType = TypeOf<typeof loggerSchema>;
+
 export const config = {
   path: 'logging',
   schema: schema.object({
     appenders: schema.mapOf(schema.string(), Appenders.configSchema, {
       defaultValue: new Map<string, AppenderConfigType>(),
     }),
-    loggers: schema.arrayOf(loggerSchema, {
-      defaultValue: [],
+    loggers: schema.mapOf(schema.string(), loggerSchema, {
+      defaultValue: new Map<string, LoggerConfigType>(),
```

Comment:
Did you do some manual testing of the behavior when specifying the loggers directly using the config file? I think all our unit & integration test are using already-converted json configs

I thought that because we are using `_.set` somewhere during yaml->json using keys such as `parent.child` in the yaml config would create a nested object (see https://github.com/elastic/kibana/issues/69592#issuecomment-648015280), and I don't any modification of that behavior in the PR.

But if I'm wrong, this is an excellent new.

Reply:
Yep you're right. I must have not tested nested context paths when I tested this manually. I don't see any great easy options for fixing this.
- We can't handle this at the `LoggingConfig` level because the nested objects create config validation errors. We could loosen the config schema with `schema.any` and then do validation in the LoggingConfig class, but I worry about having the validation error throw at a different point in time than all other config validation.
- We could customize how config is read initially. Our config reading / merging logic in `src/core/server/config/read_config.ts` has no knowledge of the shape of this data. We could have a hack in there to handle `logging.loggers` special, but I'd prefer to have a generic solution. The generic solution could take many shapes, but ideally there'd be some option that plugins and Core could use to customize how the config is constructed / read. This is probably non-trivial.
- We can postpone this change until later. I think it'd be a good change to get in before we make the new logging config GA, but we have plenty of time before 7.last to make that happen.

Reply:
This is one of the biggest flaw in the yaml format if you ask me, as this is just not supported by the official specification.

The solution/workaround the most commonly used is to use a magic/escape char to specify that the key must be considered as terminal. 

I.E in springboot, that would be (if I remember correctly)

```yaml
foo:
  "[bar.com]":
    a: b
  "[baz.com]":
    a: c
```

We could introduce a similar pattern and adapt our config parser to handle it.
Biggest Issues is that it's really error prone, as anyone editing the logger config must be aware of the specific format to use. But maybe adding some documentation/example in the `kibana.yaml` config is enough?

> We can postpone this change until later. I think it'd be a good change to get in before we make the new logging config GA

Yea, would be great to have this before GA.

> The generic solution could take many shapes, but ideally there'd be some option that plugins and Core could use to customize how the config is constructed / read. This is probably non-trivial.

Yea, even if ideal, doing this at a higher level that yaml reading/parsing is really is non-trivial. as it would require an additional level of abstraction between the yaml format and the equivalent json object.
Maybe we could expose joi's `coerce` option to `config-schema` API in some way, to allow manual / explicit coercion , but that's probably quite a big change too.

Reply:
>  Maybe we could expose joi's `coerce` option to `config-schema` API in some way, to allow manual / explicit coercion , but that's probably quite a big change too.

Yeah this is the route I was thinking with this option. Agree that it's non-trivial.

> The solution/workaround the most commonly used is to use a magic/escape char to specify that the key must be considered as terminal.
> 
> I.E in springboot, that would be (if I remember correctly)
> 
> ```yaml
> foo:
>   "[bar.com]":
>     a: b
>   "[baz.com]":
>     a: c
> ```
> 
> We could introduce a similar pattern and adapt our config parser to handle it.
> Biggest Issues is that it's really error prone, as anyone editing the logger config must be aware of the specific format to use. But maybe adding some documentation/example in the `kibana.yaml` config is enough?

This seems like an ok option, but I think it does differ from Elasticsearch. I'd like to understand how they handle this, but I'll need to dig into their code.

In terms of helping users get this right, we could document + add helpful log message when we detect errors in the part of the config where we suspect users may run into this.

I took a brief stab at implementing this. It's not complicated, however we will need to change more than just the read_config.ts file. It seems the ObjectToConfig adapters we have also call `lodash.set`.

Reply:
Oh yea, forgot about `ObjectToConfig` `get` and `set`... Even if I think that we are only using access to read 'root' sections from the config, it would probably better to also handle this syntax there.

> This seems like an ok option, but I think it does differ from Elasticsearch. I'd like to understand how they handle this, but I'll need to dig into their code.

Would be interesting to see how they are handle this, yes.

Reply:
Could this change break LP config? 
Wouldn't it be easier to implement when we get rid of the Legacy platform with another config adapter?

Reply:
> Could this change break LP config?
> Wouldn't it be easier to implement when we get rid of the Legacy platform with another config adapter?

Yeah probably so, I think I will just close this PR for the time being and revisit after we have removed legacy (7.12 or so)

Reply:
Is there an issue to track?